File: app/src/main/java/xyz/sleekstats/completist/view/CastAdapter.java
package xyz.sleekstats.completist.view;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import xyz.sleekstats.completist.databinding.CastItemBinding;
import xyz.sleekstats.completist.model.CastInfo;

//Load director and cast details for a film in a recyclerview
public class CastAdapter extends RecyclerView.Adapter<CastAdapter.CastViewHolder>{

    private List<CastInfo> castInfoList;
    private ItemClickListener mClickListener;

    public CastAdapter(List<CastInfo> castInfoList) {
        this.castInfoList = castInfoList;
    }

    @NonNull
    @Override
    public CastViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        CastItemBinding castItemBinding = CastItemBinding.inflate(layoutInflater, parent, false);
        return new CastViewHolder(castItemBinding);
    }

    //Load name and poster details for director/cast
    @Override
    public void onBindViewHolder(@NonNull CastViewHolder holder, int position) {
        holder.bind(castInfoList.get(position));
    }

    @Override
    public int getItemCount() {
        return castInfoList.size();
    }

    class CastViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final View mView;
        private final CastItemBinding castItemBinding;

        public CastViewHolder(CastItemBinding binding) {
            super(binding.getRoot());
            this.castItemBinding = binding;
            mView = castItemBinding.getRoot();
            mView.setOnClickListener(this);
        }

        public void bind(CastInfo castInfo) {
            mView.setTag(castInfo.getId());
            castItemBinding.setCast(castInfo);
            castItemBinding.executePendingBindings();
        }


        @Override
        public void onClick(View view) {
            String id = mView.getTag().toString();
            if (mClickListener != null) mClickListener.onCastClick(id);
        }
    }

    public void setCastInfoList(List<CastInfo> castInfoList) {
        this.castInfoList = castInfoList;
        notifyDataSetChanged();
    }

    public void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    //When director/actor is clicked, go to their list of films
    public interface ItemClickListener{
        void onCastClick(String castID);
    }
}

File: app/src/main/java/xyz/sleekstats/completist/view/CastAdapter.kt
package xyz.sleekstats.completist.view

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import xyz.sleekstats.completist.databinding.CastItemBinding
import xyz.sleekstats.completist.model.CastInfo
import xyz.sleekstats.completist.view.CastAdapter.CastViewHolder

//Load director and cast details for a film in a recyclerview
class CastAdapter(private var castInfoList: List<CastInfo>) : RecyclerView.Adapter<CastViewHolder>() {

    private var mClickListener: ItemClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CastViewHolder =
            CastViewHolder(CastItemBinding.inflate(LayoutInflater.from(parent.context), parent, false))

    //Load name and poster details for director/cast
    override fun onBindViewHolder(holder: CastViewHolder, position: Int) = holder.bind(castInfoList[position])

    override fun getItemCount(): Int = castInfoList.size

    inner class CastViewHolder(private val castItemBinding: CastItemBinding) : RecyclerView.ViewHolder(castItemBinding.root), View.OnClickListener {

        private val mView: View = castItemBinding.root

        init {
            mView.setOnClickListener(this)
        }

        fun bind(castInfo: CastInfo) {
            mView.tag = castInfo.id
            castItemBinding.cast = castInfo
            castItemBinding.executePendingBindings()
        }

        override fun onClick(view: View) {
            mClickListener?.onCastClick(mView.tag.toString())
        }
    }

    fun setCastInfoList(castInfoList: List<CastInfo>) {
        this.castInfoList = castInfoList
        notifyDataSetChanged()
    }

    fun setClickListener(itemClickListener: ItemClickListener?) {
        mClickListener = itemClickListener
    }

    //When director/actor is clicked, go to their list of films
    interface ItemClickListener {
        fun onCastClick(castID: String?)
    }

}